src/time-to-interactive/TimeToInteractive.ts

```diff
@@ -0,0 +1,382 @@
+    private ttiTracker: any = {};
+    private ttiResolved = false;
+
+    private fcpSupported = false;
+    private lcpSupported = false;
+    private fpsSupported = false; // Check for support for requestAnimationFrame
+
+    private COLLECTION_PERIOD = 100;
+    private REQUIRED_ACCEPTED_INTERVALS = 5;
+    private CHECK_PERIOD = 1000; // Perform another check for TTI/Visually ready
+    private VISUALLY_READY_RESOLVE_TIMEOUT = 10000;
+    private TTI_RESOLVE_TIMEOUT = 10000;
+
+    private FPS_THRESHOLD = 20 / (1000 / this.COLLECTION_PERIOD);
+    private LONG_TASK_THRESHOLD = 0;
+
+    public async computeTimeToInteractive(): Promise<number> {
+        this.initListeners();
+        return new Promise<number>((resolve, reject) => {
+            this.checkForVisualReady().then((visuallyReadyTimestamp) => {
+                this.computeTTI(visuallyReadyTimestamp).then((ttiVal) => {
+                    resolve(ttiVal);
+                });
+            });
+        });
+    }
+
+    private async checkForVisualReady(): Promise<number> {
+        // Check if visually ready
+        return new Promise<number>((resolve, reject) => {
+            let timeIntervals = 0;
+            const visuallyReadyInterval: any = setInterval(() => {
+                const isCheckTimedOut: boolean =
+                    timeIntervals * this.CHECK_PERIOD >
+                    this.VISUALLY_READY_RESOLVE_TIMEOUT;
+
+                if (this.isOkToResolveVisuallyReady(isCheckTimedOut)) {
+                    clearInterval(visuallyReadyInterval);
+
+                    this.visuallyReadyTimestamp = Math.max(
+                        this.fcpTime ? this.fcpTime : 0,
+                        this.lcpTime ? this.lcpTime : 0,
+                        this.domContentLoadedEventEnd
+                            ? this.domContentLoadedEventEnd
+                            : 0
+                    );
+                    resolve(this.visuallyReadyTimestamp);
+                } else {
+                    // No VR timestamps and check timed out, therefore stop searching
+                    if (isCheckTimedOut) {
+                        this.ttiResolved = true;
+                        this.ttiTracker = {};
+                        reject(
+                            'Insufficient visually ready timestamps to compute TTI'
+                        );
+                    }
+                }
+                timeIntervals += 1;
+            }, this.CHECK_PERIOD);
+        });
+    }
+
+    private async computeTTI(visuallyReadyTimestamp: number): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            const startBucket = Math.max(
+                this.computeTimeWindow(visuallyReadyTimestamp),
+                0
+            );
+            let currBucket = startBucket;
+            let acceptedIntervals = 0;
+            let totalIntervals = 0;
+
+            const tti: any = setInterval(() => {
```

Comment:
Nonblocking issue: setInterval should not be trusted to exactly match the expected time interval. For precision, we should be using performance.now() whenever possible

Reply:
Nevermind, it's accurate within few ms